// @ativie/ripple/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ripple = exports.RippleError = void 0;
const axios_1 = __importDefault(require("axios"));
class RippleError extends Error {
    constructor(message, statusCode, response) {
        super(message);
        this.name = 'RippleError';
        this.statusCode = statusCode;
        this.response = response;
        Object.setPrototypeOf(this, RippleError.prototype);
    }
}
exports.RippleError = RippleError;
const RETRYABLE_STATUSES = new Set([429, 500, 502, 503, 504]);
function assertString(value, name, maxLen) {
    if (typeof value !== 'string' || !value.trim()) {
        throw new RippleError(`"${name}" must be a non-empty string`);
    }
    const trimmed = value.trim();
    if (trimmed.length > maxLen) {
        throw new RippleError(`"${name}" must be ${maxLen} characters or fewer`);
    }
    return trimmed;
}
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}
class Ripple {
    constructor(options) {
        assertString(options.apiKey, 'apiKey', 500);
        assertString(options.baseUrl, 'baseUrl', 2000);
        this.retries = Math.max(0, options.retries ?? 2);
        this.client = axios_1.default.create({
            baseURL: options.baseUrl.replace(/\/+$/, ''),
            headers: {
                'Content-Type': 'application/json',
                'x-api-key': options.apiKey,
            },
            timeout: options.timeout ?? 10000,
        });
    }
    async withRetry(fn, attempt = 0) {
        try {
            return await fn();
        }
        catch (err) {
            if (!axios_1.default.isAxiosError(err))
                throw err;
            const status = err.response?.status;
            const isRetryable = !status || RETRYABLE_STATUSES.has(status);
            if (isRetryable && attempt < this.retries) {
                await sleep(Math.min(150 * 2 ** attempt, 3000));
                return this.withRetry(fn, attempt + 1);
            }
            const message = err.response?.data?.error ?? err.message;
            throw new RippleError(message, status, err.response?.data);
        }
    }
    async track(options) {
        const channel = assertString(options.channel, 'channel', 100);
        const event = assertString(options.event, 'event', 200);
        const payload = {
            channel,
            event,
            timestamp: (options.timestamp ?? new Date()).toISOString(),
        };
        if (options.user !== undefined) {
            payload.user = assertString(options.user, 'user', 100);
        }
        if (options.metadata !== undefined) {
            if (typeof options.metadata !== 'object' ||
                options.metadata === null ||
                Array.isArray(options.metadata)) {
                throw new RippleError('"metadata" must be a plain object');
            }
            payload.metadata = options.metadata;
        }
        await this.withRetry(() => this.client.post('/events', payload));
    }
    async identify(userId, metadata = {}) {
        assertString(userId, 'userId', 100);
        await this.track({
            channel: 'users',
            event: 'User Identified',
            user: userId,
            metadata,
        });
    }
}
exports.Ripple = Ripple;
exports.default = Ripple;
